precompiles/utils/src/xcm.rs

```diff
@@ -0,0 +1,354 @@
+                encoded.append(&mut network_id_to_bytes(network));
+                encoded.as_slice().into()
+            }
+            Junction::AccountKey20 { network, key } => {
+                encoded.push(3u8);
+                encoded.append(&mut key.to_vec());
+                encoded.append(&mut network_id_to_bytes(network));
+                encoded.as_slice().into()
+            }
+            Junction::PalletInstance(intance) => {
+                encoded.push(4u8);
+                encoded.append(&mut intance.to_be_bytes().to_vec());
+                encoded.as_slice().into()
+            }
+            Junction::GeneralIndex(id) => {
+                encoded.push(5u8);
+                encoded.append(&mut id.to_be_bytes().to_vec());
+                encoded.as_slice().into()
+            }
+            Junction::GeneralKey { length, data } => {
+                encoded.push(6u8);
+                encoded.push(length);
+                encoded.append(&mut data.into());
+                encoded.as_slice().into()
+            }
+            Junction::OnlyChild => {
+                encoded.push(7u8);
+                encoded.as_slice().into()
+            }
+            Junction::GlobalConsensus(network_id) => {
+                encoded.push(9u8);
+                encoded.append(&mut network_id_to_bytes(Some(network_id)));
+                encoded.as_slice().into()
+            }
+            // TODO: The only missing item here is Junciton::Plurality. This is a complex encoded
+            // type that we need to evaluate how to support
+            _ => unreachable!("Junction::Plurality not supported yet"),
+        };
+        EvmData::write(writer, encoded_bytes);
+    }
+
+    fn has_static_size() -> bool {
+        false
+    }
+}
+
+impl EvmData for Junctions {
+    fn read(reader: &mut EvmDataReader) -> EvmResult<Self> {
+        let junctions_bytes: Vec<Junction> = reader.read()?;
+        let mut junctions = Junctions::Here;
+        for item in junctions_bytes {
+            junctions
+                .push(item)
+                .map_err(|_| revert("overflow when reading junctions"))?;
+        }
+
+        Ok(junctions)
+    }
+
+    fn write(writer: &mut EvmDataWriter, value: Self) {
+        let encoded: Vec<Junction> = value.iter().map(|junction| junction.clone()).collect();
+        EvmData::write(writer, encoded);
+    }
+
+    fn has_static_size() -> bool {
+        false
+    }
+}
+
+// Cannot used derive macro since it is a foreign struct.
+impl EvmData for MultiLocation {
+    fn read(reader: &mut EvmDataReader) -> EvmResult<Self> {
+        let (parents, interior) = reader.read()?;
+        Ok(MultiLocation { parents, interior })
+    }
+
+    fn write(writer: &mut EvmDataWriter, value: Self) {
+        EvmData::write(writer, (value.parents, value.interior));
+    }
+
+    fn has_static_size() -> bool {
+        <(u8, Junctions)>::has_static_size()
+    }
+}
```

Comment:
What about unit tests for all this code?

Reply:
added.